### PR TITLE
Feat: Product page

### PR DESCRIPTION
**Pull request by MathiasDuus**

### CODE REVIEW

**Comment by MathiasDuus** [comment]
[GUI - Product page](https://trello.com/c/rbdUlHLI/27-gui-product-page)

**Review by Setanna — CHANGES_REQUESTED** [review]
Password change doesn't show snackbar updated user like other changes.

Sometimes the update will fail with following error:
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: Looking up a deactivated widget's ancestor is unsafe.

File: lib/misc/api/api_product.dart [hunk 1]
```diff
@@ -121,3 +121,37 @@ Future<List<BaseItem>> fetchUnits() async {
...
+          'name': 'Product not found',
+          'locations': [],
+        };
+      }
+    });
+  } catch (e) {
+    // Handle any exceptions that occur
+    if (kDebugMode) {
+      debugPrint('Error: $e');
+    }
```
Comment (Setanna) [inline lib/misc/api/api_product.dart:153]: should probably return future error, some like return Future.error('Error: $e');

File: lib/misc/api/api_user.dart [hunk 2]
```diff
@@ -42,3 +43,39 @@ Future<List<BaseItem>> fetchLocations() async {
...
+        debugPrint('Request failed with status: ${response.statusCode}');
+        log('Request failed with status: ${response.statusCode}');
+      }
+    }
+  } catch (e) {
+    // Handle any exceptions that occur
+    if (kDebugMode) {
+      debugPrint('Error: $e');
+      log('Error: $e');
+    }
```
Comment (Setanna) [inline lib/misc/api/api_user.dart:77]: Same as previous comment regarding returning future error. But also why log and debugprint?

File: lib/pages/product_page.dart [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                        ListView.builder(
+                          shrinkWrap: true,
+                          itemCount: product['locations'].length,
+                          itemBuilder: (BuildContext context, int index) {
+                            return ListTile(
+                              title: Text(
+                                  '${product['locations'][index]['address']}'),
+                              subtitle: Text(
+                                  'Stock: ${product['locations'][index]['pivot']['stock']} Shelf: ${product['locations'][index]['pivot']['shelf_amount']}'),
+                            );
```
Comment (Setanna) [inline lib/pages/product_page.dart:110]: Low priority but maybe reduce padding, add a divider between the locations and make the location bold?

File: lib/pages/product_page.dart [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+                        ListTile(
+                          title: Text('Category: ${product['category']}'),
+                        ),
+                        ListTile(
+                          title: Text('Barcode: ${product['barcode']}'),
+                        ),
+                        ListTile(
+                          title: Text(
+                              'Content: ${product['content']} ${product['unit']}'),
+                        ),
```
Comment (Setanna) [inline lib/pages/product_page.dart:82]: Low priority Maybe reduce space between the details lines?

File: lib/pages/product_page.dart [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+  @override
+  Widget build(BuildContext context) {
+    return FutureBuilder(
+      future: getProduct(),
+      builder: (BuildContext context, AsyncSnapshot snapshot) {
+        if (snapshot.connectionState == ConnectionState.done) {
+          return Scaffold(
+            appBar: CustomAppBar(
+              title: '${product['name']}',
```
Comment (Setanna) [inline lib/pages/product_page.dart:45]: Low priority maybe center product name for slightly better look

File: lib/pages/product_page.dart [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+              children: [
+                Padding(
+                  padding: const EdgeInsets.all(8.0),
+                  child: Card(
+                    child: ListView(
+                      shrinkWrap: true,
+                      children: [
+                        const ListTile(
+                          title: Text(
+                            'Details:',
```
Comment (Setanna) [inline lib/pages/product_page.dart:66]: Low priority could center details title for slightly better look

File: lib/pages/product_page.dart [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+                  ),
+                ),
+                Padding(
+                  padding: const EdgeInsets.all(8.0),
+                  child: Card(
+                    child: Column(
+                      children: [
+                        const ListTile(
+                          title: Text(
+                            'Locations:',
```
Comment (Setanna) [inline lib/pages/product_page.dart:94]: Low priority could center locations title for slightly better look

**Review by Setanna — CHANGES_REQUESTED** [review]

File: lib/misc/api/api_product.dart (hunk 1 above)
Comment (MathiasDuus) [reply lib/misc/api/api_product.dart:153]: Done

File: lib/misc/api/api_user.dart (hunk 2 above)
Comment (MathiasDuus) [reply lib/misc/api/api_user.dart]: Because if it fails, it should just return false.

File: lib/pages/product_page.dart (hunk 4 above)
Comment (MathiasDuus) [reply lib/pages/product_page.dart]: Done

**Review by Setanna — APPROVED** [review]

**Pull request merged**